pallets/runtime/tests/src/transaction_ordering_test.rs

```diff
@@ -0,0 +1,1056 @@
+                alice_creds.clone(),
+                dave_creds.clone(),
+                charlie_creds.clone(),
+                charlie_secret_account.clone(),
+                alice_init_balance - alice_sent_amount_1001,
+                dave_init_balance + dave_received_amount_1001,
+                true, // Validation failure expected.
+            );
+
+            // Approve and process tx:1000.
+            finalize_transaction(
+                instruction_counter1000,
+                alice_creds.clone(),
+                bob_creds.clone(),
+                charlie_creds.clone(),
+                charlie_secret_account.clone(),
+                alice_init_balance - alice_sent_amount_1000,
+                bob_init_balance + bob_received_amount_1000,
+                false,
+            );
+        });
+}
+
+#[test]
+fn mercat_whitepaper_scenario1() {
+    ExtBuilder::default()
+        .cdd_providers(vec![AccountKeyring::Eve.public()])
+        .set_max_legs_allowed(500)
+        .build()
+        .execute_with(|| {
+            // Setting:
+            //   - Alice is the token issuer, and has 10 assets in her supply.
+            //   - Bob has a normal account.
+            //   - Dave has a normal account.
+            //   - Charlie is the mediator.
+            //   - Eve is the CDD provider.
+            let (
+                alice_creds,
+                alice_secret_account,
+                alice_init_balance,
+                charlie_creds,
+                charlie_secret_account,
+            ) = chain_set_up(90u128);
+
+            let (bob_secret_account, bob_creds, _) =
+                create_investor_account(2u32, AccountKeyring::Bob);
+
+            let (dave_secret_account, dave_creds, dave_init_balance) =
+                create_investor_account(3u32, AccountKeyring::Dave);
+
+            // Alice, the token issuer, sends 10 tokens to Dave so he has something in his account.
+            let (instruction_counter999, alice_sent_amount_999, dave_received_amount_999) =
+                initialize_transaction(
+                    alice_secret_account.clone(),
+                    alice_creds.clone(),
+                    alice_init_balance.clone(),
+                    dave_secret_account.clone(),
+                    dave_creds.clone(),
+                    charlie_creds.clone(),
+                    1u32, // tx_id.
+                    10,
+                );
+            finalize_transaction(
+                instruction_counter999,
+                alice_creds.clone(),
+                dave_creds.clone(),
+                charlie_creds.clone(),
+                charlie_secret_account.clone(),
+                alice_init_balance - alice_sent_amount_999,
+                dave_init_balance + dave_received_amount_999,
+                false,
+            );
+            // Reset Dave's pending state.
+            assert_ok!(ConfidentialAsset::update_account_balance_checkpoint(
```

Comment:
whose responsibility is it to call this function? Account holder or the chain? I would assume that it is the chain's responsibility and therefore I did not expect to see this call in the test. I expected this to be handled under the hood.

Reply:
It's the responsibility of the user/wallet to call this function. Chain keeps updating the _accounts balance_ as the transactions are settled, but it's the wallet that should control _pending state_'s transition. What we are doing on the chain side by keeping all the accumulators is to recreate the pending state that was used on the wallet side. 

Here in this unit test, Alice (asset issuer) transfers some assets to Dave, but Dave has to switch his checkpoint to this new balance so he can spend them.